R/conclass.R

```diff
@@ -354,26 +354,33 @@ Conos <- setRefClass(
         plotDEGenes(de.genes, samples, groups=groups, n.genes.to.show=n.genes.to.show, inner.clustering=inner.clustering)
       }
 
-      if (append.specifisity.metrics) {
+      if (append.specificity.metrics) {
         lapply.func <- if (verbose) function(...) pbapply::pblapply(..., cl=n.jobs) else function(...) papply(..., n.cores=n.jobs)
-        if (verbose) cat("Estimating specifisity metrics\n")
+        if (verbose) cat("Estimating specificity metrics\n")
 
         cm.merged <- lapply(samples, getRawCountMatrix, transposed=T) %>% mergeCountMatrices(transposed=T)
+        groups.clean <- groups[!is.na(groups)]
+        dif <- intersect(rownames(cm.merged),names(groups.clean))
 
-        if (length(intersect(rownames(cm.merged), names(groups))) != nrow(cm.merged))
-          stop("`groups` must contain values for all cells in the samples")
-
-        if (nrow(cm.merged) < length(groups)) {
-          groups %<>% .[rownames(cm.merged)]
+        if(nrow(cm.merged) != length(groups.clean) | length(dif) != nrow(cm.merged)) {
+          if(nrow(cm.merged) < length(groups.clean)) {
+            if(verbose) message("'groups' is defined for ",length(groups.clean)-nrow(cm.merged)," unknown cells.")
+            groups.clean %<>% subset(rownames %in% rownames(cm.merged))
+          }
+          if(length(dif) < nrow(cm.merged)) {
+            if(verbose) message("Excluding ",nrow(cm.merged)-length(dif)," cells missing from 'groups'.")
+            cm.merged %<>% .[rownames(.) %in% dif,]
+          }
         }
 
-        de.genes %<>% names() %>% setNames(., .) %>%
-          lapply.func(function(n) appendSpecifisityMetricsToDE(de.genes[[n]], groups, n, p2.counts=cm.merged, append.auc=append.auc))
+        de.genes.clean <- de.genes %<>% lapply(function(x) x %<>% subset(complete.cases(.)))
+        de.genes.clean %<>% names() %>% setNames(., .) %>%
+          lapply.func(function(n) appendSpecificityMetricsToDE(de.genes.clean[[n]], groups.clean, n, p2.counts=cm.merged, append.auc=append.auc))
       }
 
       if (verbose) cat("All done!\n")
 
-      return(de.genes)
+      return(de.genes.clean)
```

Comment:
You create `de.genes.clean` only with `append.specificity.metrics=T`. So the code would fail without it. I don't see particular need in storing it as a separate variable, so it should be fine to rename everything to `de.genes`

Reply:
Changed back